### PR TITLE
support PKTINFO cmsgs for Linux/macOS/iOS

### PR DESCRIPTION
**Pull request by mcginty**
The libc changes have been merged (https://github.com/rust-lang/libc/pull/980) now, so looks like it's finally time to submit a PR for this :).

### CODE REVIEW

File: src/sys/socket/mod.rs [hunk 1]
```diff
@@ -414,6 +414,14 @@ impl<'a> Iterator for CmsgIterator<'a> {
                 Some(ControlMessage::ScmTimestamp(
                     &*(cmsg_data.as_ptr() as *const _)))
             },
+            #[cfg(any(target_os = "linux", target_os = "android", target_os = "macos", target_os = "ios"))]
+            (libc::IPPROTO_IPV6, libc::IPV6_PKTINFO) => unsafe {
```
Comment (asomers) [inline src/sys/socket/mod.rs]: FreeBSD also defines `IPV6_PKTINFO`.  I haven't checked on any of the other BSDs.

**Comment by mcginty** [comment]
Agreed about tests, it is tricky code. Will do.

Yeah the other BSDs will take longer because I only merged macOS/Linux
libc constants so far just for the sake of speed. Would you like to hold
off on this whole thing until then? Or would you accept separate PRs
after I get the rest merged in libc?
On Sun, Apr 29, 2018, at 4:27 AM, Alan Somers wrote:
> *@asomers* requested changes on this pull request.


> Please enable IPV6_PKTINFO on FreeBSD, and check the other BSDs to see
> if any of them support it.  Also, can you write a test for this?  The
> cmsg code is devilishly difficult; it would be great to have some test
> coverage for every supported message type.> 
> In src/sys/socket/mod.rs[1]:


> > @@ -414,6 +414,14 @@ impl<'a> Iterator for CmsgIterator<'a> {
> > Some(ControlMessage::ScmTimestamp( &*(cmsg_data.as_ptr() as *const
> > _))) }, +            #[cfg(any(target_os = "linux", target_os =
> > "android", target_os = "macos", target_os = "ios"))] +
> > (libc::IPPROTO_IPV6, libc::IPV6_PKTINFO) => unsafe {
>> FreeBSD also defines IPV6_PKTINFO.  I haven't checked on any of the
> other BSDs.> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[2], or mute the thread[3].> 


Links:

  1. https://github.com/nix-rust/nix/pull/891#discussion_r184864071
  2. https://github.com/nix-rust/nix/pull/891#pullrequestreview-116147810

**Comment by asomers** [comment]
I'll wait for the libc PR.  libc usually merges stuff quickly.

**Comment by mcginty** [comment]
new libc PR submitted: https://github.com/rust-lang/libc/pull/983

**Comment by sjtakada** [comment]
I'm curious if this is still in progress, because I also need PKTINFO.

File: src/sys/socket/mod.rs [hunk 2]
```diff
@@ -414,6 +414,14 @@ impl<'a> Iterator for CmsgIterator<'a> {
                 Some(ControlMessage::ScmTimestamp(
                     &*(cmsg_data.as_ptr() as *const _)))
             },
+            #[cfg(any(target_os = "linux", target_os = "android", target_os = "macos", target_os = "ios"))]
```
Comment (Susurrus) [inline src/sys/socket/mod.rs]: These lists should be alphabetized (android, ios, linux, macos). This applies to all of your `#[cfg` attributes.

File: CHANGELOG.md [hunk 3]
```diff
@@ -33,6 +33,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   ([#857](https://github.com/nix-rust/nix/pull/857))
 - Added `request_code_write_int!` on FreeBSD/DragonFlyBSD
   ([#833](https://github.com/nix-rust/nix/pull/833))
+- Added PKTINFO cmsg support on Linux/macOS/iOS.
```
Comment (Susurrus) [inline CHANGELOG.md]: Change to "Added \`PKTINFO\` support for \`cmsg\` on ... platforms" (note the backticks)

**Comment by pusateri** [comment]
Thanks @mcinty but I don't need credit. Just would like to use this feature. I can push my changes back to your branch if you give me permission. I'm just not familiar with how to update someone else's pull request.

**Comment by pusateri** [comment]
I submitted a new pull request that updates the changes, adds tests without using pre-configured loopback addresses, adds IPv6_PKTINFO for FreeBSD, and sorts the cfg options. See #990.

**Comment by mcginty** [comment]
Sorry for the delay @pusateri. Closing in favor of #990.

**Pull request closed without merging**